{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# DSS Model Mapping"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "from datetime import date"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load DSS Model from Excel"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "RAW_DSS_MAP_EXCEL = Path(\"../data/input/CRDC DST Cross-Model Mapping.xlsx\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\ProgramData\\Anaconda3\\envs\\bento_test_3_8\\lib\\site-packages\\openpyxl\\worksheet\\_reader.py:312: UserWarning: Data Validation extension is not supported and will be removed\n",
      "  warn(msg)\n"
     ]
    }
   ],
   "source": [
    "raw_df = pd.read_excel(\n",
    "    RAW_DSS_MAP_EXCEL,\n",
    "    sheet_name=\"CRDC DST Cross-Model Mapping\",\n",
    "    header=2,\n",
    "    nrows=1529,\n",
    "    usecols=[\"DST Data Element Name\", \"Compiled Data Element Names\"]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>DST Data Element Name</th>\n",
       "      <th>Compiled Data Element Names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Subject Identifier</td>\n",
       "      <td>CTDC.case.case_id\\nGDC.Case.id\\nICDC.case.case...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Specimen Type</td>\n",
       "      <td>GDC.Sample.sample_type\\nICDC.sample.physical_s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Subject Ethnicity*</td>\n",
       "      <td>CTDC.case.ethnicity\\nGDC.Demographic.ethnicity...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Subject Race</td>\n",
       "      <td>CTDC.case.race\\nGDC.Demographic.race\\nPDC.Demo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC.Diagnosis.site_of_resection_or_biopsy\\nIDC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>Tumor Grade</td>\n",
       "      <td>PDC.Diagnosis.who_nte_grade</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>Tumor Grade</td>\n",
       "      <td>PDC.Diagnosis.gleason_grade_tertiary</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC.Diagnosis.tumor_depth\\nPDC.Diagnosis.tumor...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>Disease Clinical Stage</td>\n",
       "      <td>PDC.Diagnosis.figo_staging_edition_year</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>Exposure Type</td>\n",
       "      <td>PDC.Exposure.exposure_type</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1529 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       DST Data Element Name  \\\n",
       "0         Subject Identifier   \n",
       "1              Specimen Type   \n",
       "2         Subject Ethnicity*   \n",
       "3               Subject Race   \n",
       "4                        NaN   \n",
       "...                      ...   \n",
       "1524             Tumor Grade   \n",
       "1525             Tumor Grade   \n",
       "1526                     NaN   \n",
       "1527  Disease Clinical Stage   \n",
       "1528           Exposure Type   \n",
       "\n",
       "                            Compiled Data Element Names  \n",
       "0     CTDC.case.case_id\\nGDC.Case.id\\nICDC.case.case...  \n",
       "1     GDC.Sample.sample_type\\nICDC.sample.physical_s...  \n",
       "2     CTDC.case.ethnicity\\nGDC.Demographic.ethnicity...  \n",
       "3     CTDC.case.race\\nGDC.Demographic.race\\nPDC.Demo...  \n",
       "4     GDC.Diagnosis.site_of_resection_or_biopsy\\nIDC...  \n",
       "...                                                 ...  \n",
       "1524                        PDC.Diagnosis.who_nte_grade  \n",
       "1525               PDC.Diagnosis.gleason_grade_tertiary  \n",
       "1526  GDC.Diagnosis.tumor_depth\\nPDC.Diagnosis.tumor...  \n",
       "1527            PDC.Diagnosis.figo_staging_edition_year  \n",
       "1528                         PDC.Exposure.exposure_type  \n",
       "\n",
       "[1529 rows x 2 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_df"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Separate DSS & Other Models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [],
   "source": [
    "# model for dst nodes\n",
    "df_dss = pd.DataFrame(np.repeat(\"DSS\", len(raw_df)))\n",
    "\n",
    "df_dss = pd.concat(\n",
    "    [df_dss, raw_df[\"DST Data Element Name\"].str.split(\" \", 1, expand=True)],\n",
    "    axis=1\n",
    ")\n",
    "df_dss.columns = [\"ent_1_model\", \"ent_1_extra_handles\", \"ent_1_handle\"]\n",
    "df_dss = df_dss[[\"ent_1_model\", \"ent_1_handle\", \"ent_1_extra_handles\"]]\n",
    "# df_dst[\"ent_1_extra_handles\"] = [[l] for l in df_dst[\"ent_1_extra_handles\"]] # listify extra handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Race</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Exposure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1529 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model    ent_1_handle ent_1_extra_handles\n",
       "0            DSS      Identifier             Subject\n",
       "1            DSS            Type            Specimen\n",
       "2            DSS      Ethnicity*             Subject\n",
       "3            DSS            Race             Subject\n",
       "4            DSS             NaN                 NaN\n",
       "...          ...             ...                 ...\n",
       "1524         DSS           Grade               Tumor\n",
       "1525         DSS           Grade               Tumor\n",
       "1526         DSS             NaN                 NaN\n",
       "1527         DSS  Clinical Stage             Disease\n",
       "1528         DSS            Type            Exposure\n",
       "\n",
       "[1529 rows x 3 columns]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dss"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model_Data_Element_Name_0</th>\n",
       "      <th>Model_Data_Element_Name_1</th>\n",
       "      <th>Model_Data_Element_Name_2</th>\n",
       "      <th>Model_Data_Element_Name_3</th>\n",
       "      <th>Model_Data_Element_Name_4</th>\n",
       "      <th>Model_Data_Element_Name_5</th>\n",
       "      <th>Model_Data_Element_Name_6</th>\n",
       "      <th>Model_Data_Element_Name_7</th>\n",
       "      <th>Model_Data_Element_Name_8</th>\n",
       "      <th>Model_Data_Element_Name_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTDC.case.case_id</td>\n",
       "      <td>GDC.Case.id</td>\n",
       "      <td>ICDC.case.case_id</td>\n",
       "      <td>IDC.DICOM.Patient Module.Patient ID</td>\n",
       "      <td>PDC.Case.case_id</td>\n",
       "      <td>CDS.Participant.subject_id</td>\n",
       "      <td>CDA.Patient.dct:identifier</td>\n",
       "      <td></td>\n",
       "      <td>C2M2.biosample.local_id</td>\n",
       "      <td>mCODE.Cancer Patient Profile.Identifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>GDC.Sample.sample_type</td>\n",
       "      <td>ICDC.sample.physical_sample_type</td>\n",
       "      <td>PDC.Sample.sample_type</td>\n",
       "      <td>CDS.Sample Information.sample_type</td>\n",
       "      <td>CDA.Specimen.source_material_type</td>\n",
       "      <td>HTAN.Molecular Test.biospecimen_type</td>\n",
       "      <td>C2M2.biosample.assay_type</td>\n",
       "      <td>mCODE.Genomic Specimen Profile.Type</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTDC.case.ethnicity</td>\n",
       "      <td>GDC.Demographic.ethnicity</td>\n",
       "      <td>IDC.DICOM.Patient Module.Patient's Ethnic Group</td>\n",
       "      <td>PDC.Demographic.ethnicity</td>\n",
       "      <td>CDS.Participant.ethnicity</td>\n",
       "      <td>CDA.Patient.ethnicity</td>\n",
       "      <td>HTAN.Demographics.Ethnicity</td>\n",
       "      <td>C2M2.subject.ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CTDC.case.race</td>\n",
       "      <td>GDC.Demographic.race</td>\n",
       "      <td>PDC.Demographic.race</td>\n",
       "      <td>CDS.Participant.race</td>\n",
       "      <td>CDA.Patient.race</td>\n",
       "      <td>HTAN.Demographics.Race</td>\n",
       "      <td>C2M2.subject_race.race</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>GDC.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>IDC.TCIA.tcia_tumorLocation</td>\n",
       "      <td>PDC.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>CDS.Additional Diagnosis Information.site_of_r...</td>\n",
       "      <td>HTAN.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>PDC.Diagnosis.who_nte_grade</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>PDC.Diagnosis.gleason_grade_tertiary</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>GDC.Diagnosis.tumor_depth</td>\n",
       "      <td>PDC.Diagnosis.tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>PDC.Diagnosis.figo_staging_edition_year</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>PDC.Exposure.exposure_type</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1529 rows × 10 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      Model_Data_Element_Name_0  \\\n",
       "0                             CTDC.case.case_id   \n",
       "1                        GDC.Sample.sample_type   \n",
       "2                           CTDC.case.ethnicity   \n",
       "3                                CTDC.case.race   \n",
       "4     GDC.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                         ...   \n",
       "1524                PDC.Diagnosis.who_nte_grade   \n",
       "1525       PDC.Diagnosis.gleason_grade_tertiary   \n",
       "1526                  GDC.Diagnosis.tumor_depth   \n",
       "1527    PDC.Diagnosis.figo_staging_edition_year   \n",
       "1528                 PDC.Exposure.exposure_type   \n",
       "\n",
       "              Model_Data_Element_Name_1  \\\n",
       "0                           GDC.Case.id   \n",
       "1     ICDC.sample.physical_sample_type    \n",
       "2             GDC.Demographic.ethnicity   \n",
       "3                  GDC.Demographic.race   \n",
       "4           IDC.TCIA.tcia_tumorLocation   \n",
       "...                                 ...   \n",
       "1524                               None   \n",
       "1525                               None   \n",
       "1526          PDC.Diagnosis.tumor_depth   \n",
       "1527                               None   \n",
       "1528                               None   \n",
       "\n",
       "                            Model_Data_Element_Name_2  \\\n",
       "0                                   ICDC.case.case_id   \n",
       "1                              PDC.Sample.sample_type   \n",
       "2     IDC.DICOM.Patient Module.Patient's Ethnic Group   \n",
       "3                                PDC.Demographic.race   \n",
       "4           PDC.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                               ...   \n",
       "1524                                             None   \n",
       "1525                                             None   \n",
       "1526                                             None   \n",
       "1527                                             None   \n",
       "1528                                             None   \n",
       "\n",
       "                              Model_Data_Element_Name_3  \\\n",
       "0                   IDC.DICOM.Patient Module.Patient ID   \n",
       "1                    CDS.Sample Information.sample_type   \n",
       "2                             PDC.Demographic.ethnicity   \n",
       "3                                  CDS.Participant.race   \n",
       "4     CDS.Additional Diagnosis Information.site_of_r...   \n",
       "...                                                 ...   \n",
       "1524                                               None   \n",
       "1525                                               None   \n",
       "1526                                               None   \n",
       "1527                                               None   \n",
       "1528                                               None   \n",
       "\n",
       "                       Model_Data_Element_Name_4  \\\n",
       "0                               PDC.Case.case_id   \n",
       "1              CDA.Specimen.source_material_type   \n",
       "2                      CDS.Participant.ethnicity   \n",
       "3                               CDA.Patient.race   \n",
       "4     HTAN.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                          ...   \n",
       "1524                                        None   \n",
       "1525                                        None   \n",
       "1526                                        None   \n",
       "1527                                        None   \n",
       "1528                                        None   \n",
       "\n",
       "                 Model_Data_Element_Name_5    Model_Data_Element_Name_6  \\\n",
       "0               CDS.Participant.subject_id   CDA.Patient.dct:identifier   \n",
       "1     HTAN.Molecular Test.biospecimen_type    C2M2.biosample.assay_type   \n",
       "2                    CDA.Patient.ethnicity  HTAN.Demographics.Ethnicity   \n",
       "3                   HTAN.Demographics.Race       C2M2.subject_race.race   \n",
       "4                                     None                         None   \n",
       "...                                    ...                          ...   \n",
       "1524                                  None                         None   \n",
       "1525                                  None                         None   \n",
       "1526                                  None                         None   \n",
       "1527                                  None                         None   \n",
       "1528                                  None                         None   \n",
       "\n",
       "                Model_Data_Element_Name_7 Model_Data_Element_Name_8  \\\n",
       "0                                           C2M2.biosample.local_id   \n",
       "1     mCODE.Genomic Specimen Profile.Type                      None   \n",
       "2                  C2M2.subject.ethnicity                      None   \n",
       "3                                    None                      None   \n",
       "4                                    None                      None   \n",
       "...                                   ...                       ...   \n",
       "1524                                 None                      None   \n",
       "1525                                 None                      None   \n",
       "1526                                 None                      None   \n",
       "1527                                 None                      None   \n",
       "1528                                 None                      None   \n",
       "\n",
       "                    Model_Data_Element_Name_9  \n",
       "0     mCODE.Cancer Patient Profile.Identifier  \n",
       "1                                        None  \n",
       "2                                        None  \n",
       "3                                        None  \n",
       "4                                        None  \n",
       "...                                       ...  \n",
       "1524                                     None  \n",
       "1525                                     None  \n",
       "1526                                     None  \n",
       "1527                                     None  \n",
       "1528                                     None  \n",
       "\n",
       "[1529 rows x 10 columns]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models = raw_df[\"Compiled Data Element Names\"].str.strip().str.split(\n",
    "    '\\n', \n",
    "    expand=True\n",
    ").add_prefix(\"Model_Data_Element_Name_\")\n",
    "\n",
    "df_models"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>Model_Data_Element_Name_0</th>\n",
       "      <th>Model_Data_Element_Name_1</th>\n",
       "      <th>Model_Data_Element_Name_2</th>\n",
       "      <th>Model_Data_Element_Name_3</th>\n",
       "      <th>Model_Data_Element_Name_4</th>\n",
       "      <th>Model_Data_Element_Name_5</th>\n",
       "      <th>Model_Data_Element_Name_6</th>\n",
       "      <th>Model_Data_Element_Name_7</th>\n",
       "      <th>Model_Data_Element_Name_8</th>\n",
       "      <th>Model_Data_Element_Name_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC.case.case_id</td>\n",
       "      <td>GDC.Case.id</td>\n",
       "      <td>ICDC.case.case_id</td>\n",
       "      <td>IDC.DICOM.Patient Module.Patient ID</td>\n",
       "      <td>PDC.Case.case_id</td>\n",
       "      <td>CDS.Participant.subject_id</td>\n",
       "      <td>CDA.Patient.dct:identifier</td>\n",
       "      <td></td>\n",
       "      <td>C2M2.biosample.local_id</td>\n",
       "      <td>mCODE.Cancer Patient Profile.Identifier</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>GDC.Sample.sample_type</td>\n",
       "      <td>ICDC.sample.physical_sample_type</td>\n",
       "      <td>PDC.Sample.sample_type</td>\n",
       "      <td>CDS.Sample Information.sample_type</td>\n",
       "      <td>CDA.Specimen.source_material_type</td>\n",
       "      <td>HTAN.Molecular Test.biospecimen_type</td>\n",
       "      <td>C2M2.biosample.assay_type</td>\n",
       "      <td>mCODE.Genomic Specimen Profile.Type</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC.case.ethnicity</td>\n",
       "      <td>GDC.Demographic.ethnicity</td>\n",
       "      <td>IDC.DICOM.Patient Module.Patient's Ethnic Group</td>\n",
       "      <td>PDC.Demographic.ethnicity</td>\n",
       "      <td>CDS.Participant.ethnicity</td>\n",
       "      <td>CDA.Patient.ethnicity</td>\n",
       "      <td>HTAN.Demographics.Ethnicity</td>\n",
       "      <td>C2M2.subject.ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Race</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC.case.race</td>\n",
       "      <td>GDC.Demographic.race</td>\n",
       "      <td>PDC.Demographic.race</td>\n",
       "      <td>CDS.Participant.race</td>\n",
       "      <td>CDA.Patient.race</td>\n",
       "      <td>HTAN.Demographics.Race</td>\n",
       "      <td>C2M2.subject_race.race</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>IDC.TCIA.tcia_tumorLocation</td>\n",
       "      <td>PDC.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>CDS.Additional Diagnosis Information.site_of_r...</td>\n",
       "      <td>HTAN.Diagnosis.site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC.Diagnosis.who_nte_grade</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC.Diagnosis.gleason_grade_tertiary</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC.Diagnosis.tumor_depth</td>\n",
       "      <td>PDC.Diagnosis.tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PDC.Diagnosis.figo_staging_edition_year</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Exposure</td>\n",
       "      <td>PDC.Exposure.exposure_type</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1529 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model    ent_1_handle ent_1_extra_handles  \\\n",
       "0            DSS      Identifier             Subject   \n",
       "1            DSS            Type            Specimen   \n",
       "2            DSS      Ethnicity*             Subject   \n",
       "3            DSS            Race             Subject   \n",
       "4            DSS             NaN                 NaN   \n",
       "...          ...             ...                 ...   \n",
       "1524         DSS           Grade               Tumor   \n",
       "1525         DSS           Grade               Tumor   \n",
       "1526         DSS             NaN                 NaN   \n",
       "1527         DSS  Clinical Stage             Disease   \n",
       "1528         DSS            Type            Exposure   \n",
       "\n",
       "                      Model_Data_Element_Name_0  \\\n",
       "0                             CTDC.case.case_id   \n",
       "1                        GDC.Sample.sample_type   \n",
       "2                           CTDC.case.ethnicity   \n",
       "3                                CTDC.case.race   \n",
       "4     GDC.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                         ...   \n",
       "1524                PDC.Diagnosis.who_nte_grade   \n",
       "1525       PDC.Diagnosis.gleason_grade_tertiary   \n",
       "1526                  GDC.Diagnosis.tumor_depth   \n",
       "1527    PDC.Diagnosis.figo_staging_edition_year   \n",
       "1528                 PDC.Exposure.exposure_type   \n",
       "\n",
       "              Model_Data_Element_Name_1  \\\n",
       "0                           GDC.Case.id   \n",
       "1     ICDC.sample.physical_sample_type    \n",
       "2             GDC.Demographic.ethnicity   \n",
       "3                  GDC.Demographic.race   \n",
       "4           IDC.TCIA.tcia_tumorLocation   \n",
       "...                                 ...   \n",
       "1524                               None   \n",
       "1525                               None   \n",
       "1526          PDC.Diagnosis.tumor_depth   \n",
       "1527                               None   \n",
       "1528                               None   \n",
       "\n",
       "                            Model_Data_Element_Name_2  \\\n",
       "0                                   ICDC.case.case_id   \n",
       "1                              PDC.Sample.sample_type   \n",
       "2     IDC.DICOM.Patient Module.Patient's Ethnic Group   \n",
       "3                                PDC.Demographic.race   \n",
       "4           PDC.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                               ...   \n",
       "1524                                             None   \n",
       "1525                                             None   \n",
       "1526                                             None   \n",
       "1527                                             None   \n",
       "1528                                             None   \n",
       "\n",
       "                              Model_Data_Element_Name_3  \\\n",
       "0                   IDC.DICOM.Patient Module.Patient ID   \n",
       "1                    CDS.Sample Information.sample_type   \n",
       "2                             PDC.Demographic.ethnicity   \n",
       "3                                  CDS.Participant.race   \n",
       "4     CDS.Additional Diagnosis Information.site_of_r...   \n",
       "...                                                 ...   \n",
       "1524                                               None   \n",
       "1525                                               None   \n",
       "1526                                               None   \n",
       "1527                                               None   \n",
       "1528                                               None   \n",
       "\n",
       "                       Model_Data_Element_Name_4  \\\n",
       "0                               PDC.Case.case_id   \n",
       "1              CDA.Specimen.source_material_type   \n",
       "2                      CDS.Participant.ethnicity   \n",
       "3                               CDA.Patient.race   \n",
       "4     HTAN.Diagnosis.site_of_resection_or_biopsy   \n",
       "...                                          ...   \n",
       "1524                                        None   \n",
       "1525                                        None   \n",
       "1526                                        None   \n",
       "1527                                        None   \n",
       "1528                                        None   \n",
       "\n",
       "                 Model_Data_Element_Name_5    Model_Data_Element_Name_6  \\\n",
       "0               CDS.Participant.subject_id   CDA.Patient.dct:identifier   \n",
       "1     HTAN.Molecular Test.biospecimen_type    C2M2.biosample.assay_type   \n",
       "2                    CDA.Patient.ethnicity  HTAN.Demographics.Ethnicity   \n",
       "3                   HTAN.Demographics.Race       C2M2.subject_race.race   \n",
       "4                                     None                         None   \n",
       "...                                    ...                          ...   \n",
       "1524                                  None                         None   \n",
       "1525                                  None                         None   \n",
       "1526                                  None                         None   \n",
       "1527                                  None                         None   \n",
       "1528                                  None                         None   \n",
       "\n",
       "                Model_Data_Element_Name_7 Model_Data_Element_Name_8  \\\n",
       "0                                           C2M2.biosample.local_id   \n",
       "1     mCODE.Genomic Specimen Profile.Type                      None   \n",
       "2                  C2M2.subject.ethnicity                      None   \n",
       "3                                    None                      None   \n",
       "4                                    None                      None   \n",
       "...                                   ...                       ...   \n",
       "1524                                 None                      None   \n",
       "1525                                 None                      None   \n",
       "1526                                 None                      None   \n",
       "1527                                 None                      None   \n",
       "1528                                 None                      None   \n",
       "\n",
       "                    Model_Data_Element_Name_9  \n",
       "0     mCODE.Cancer Patient Profile.Identifier  \n",
       "1                                        None  \n",
       "2                                        None  \n",
       "3                                        None  \n",
       "4                                        None  \n",
       "...                                       ...  \n",
       "1524                                     None  \n",
       "1525                                     None  \n",
       "1526                                     None  \n",
       "1527                                     None  \n",
       "1528                                     None  \n",
       "\n",
       "[1529 rows x 13 columns]"
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_dss.join(df_models)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop duplicates in whole set of mappings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1522 1522\n"
     ]
    }
   ],
   "source": [
    "# check for duplicates in whole df, drop 2nd occurance\n",
    "df_both = df_dss.join(df_models)\n",
    "df_both.drop_duplicates(keep=\"first\", inplace=True)\n",
    "dss_df = df_both[[\"ent_1_model\", \"ent_1_handle\", \"ent_1_extra_handles\"]]\n",
    "df_models = df_both.drop([\"ent_1_model\", \"ent_1_handle\", \"ent_1_extra_handles\"], axis=1)\n",
    "print(len(dss_df), len(df_models))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Model_Data_Element_Name_0    1520\n",
       "Model_Data_Element_Name_1     394\n",
       "Model_Data_Element_Name_2     243\n",
       "Model_Data_Element_Name_3      66\n",
       "Model_Data_Element_Name_4      32\n",
       "Model_Data_Element_Name_5      18\n",
       "Model_Data_Element_Name_6      10\n",
       "Model_Data_Element_Name_7       4\n",
       "Model_Data_Element_Name_8       2\n",
       "Model_Data_Element_Name_9       1\n",
       "dtype: int64"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0 10\n",
      "48 9\n"
     ]
    }
   ],
   "source": [
    "# look at row with 9 \"mappings\"\n",
    "for i, x in enumerate(df_models.count(axis=1)):\n",
    "    if x >= 9:\n",
    "        print(i, x)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Model_Data_Element_Name_0</th>\n",
       "      <th>Model_Data_Element_Name_1</th>\n",
       "      <th>Model_Data_Element_Name_2</th>\n",
       "      <th>Model_Data_Element_Name_3</th>\n",
       "      <th>Model_Data_Element_Name_4</th>\n",
       "      <th>Model_Data_Element_Name_5</th>\n",
       "      <th>Model_Data_Element_Name_6</th>\n",
       "      <th>Model_Data_Element_Name_7</th>\n",
       "      <th>Model_Data_Element_Name_8</th>\n",
       "      <th>Model_Data_Element_Name_9</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>GDC.Sample.biospecimen_anatomic_site</td>\n",
       "      <td>ICDC.sample.sample_site</td>\n",
       "      <td>IDC.DICOM.General Series Module.BodyPartExamined</td>\n",
       "      <td>PDC.Sample.biospecimen_anatomic_site</td>\n",
       "      <td>CDS.Additional Sample Information.sample_anato...</td>\n",
       "      <td>CDA.Specimen.anatomical_site</td>\n",
       "      <td></td>\n",
       "      <td>C2M2.biosample.anatomy</td>\n",
       "      <td>mCODE.Genomic Specimen Profile.Collection &gt; Bo...</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               Model_Data_Element_Name_0 Model_Data_Element_Name_1  \\\n",
       "48  GDC.Sample.biospecimen_anatomic_site   ICDC.sample.sample_site   \n",
       "\n",
       "                           Model_Data_Element_Name_2  \\\n",
       "48  IDC.DICOM.General Series Module.BodyPartExamined   \n",
       "\n",
       "               Model_Data_Element_Name_3  \\\n",
       "48  PDC.Sample.biospecimen_anatomic_site   \n",
       "\n",
       "                            Model_Data_Element_Name_4  \\\n",
       "48  CDS.Additional Sample Information.sample_anato...   \n",
       "\n",
       "       Model_Data_Element_Name_5 Model_Data_Element_Name_6  \\\n",
       "48  CDA.Specimen.anatomical_site                             \n",
       "\n",
       "   Model_Data_Element_Name_7  \\\n",
       "48    C2M2.biosample.anatomy   \n",
       "\n",
       "                            Model_Data_Element_Name_8  \\\n",
       "48  mCODE.Genomic Specimen Profile.Collection > Bo...   \n",
       "\n",
       "   Model_Data_Element_Name_9  \n",
       "48                      None  "
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models.iloc[[48]]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Stack rows of matches into one column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_models_stack = pd.DataFrame(df_models.stack())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">0</th>\n",
       "      <th>Model_Data_Element_Name_0</th>\n",
       "      <td>CTDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model_Data_Element_Name_1</th>\n",
       "      <td>GDC.Case.id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model_Data_Element_Name_2</th>\n",
       "      <td>ICDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model_Data_Element_Name_3</th>\n",
       "      <td>IDC.DICOM.Patient Module.Patient ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Model_Data_Element_Name_4</th>\n",
       "      <td>PDC.Case.case_id</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                               0\n",
       "0 Model_Data_Element_Name_0                    CTDC.case.case_id\n",
       "  Model_Data_Element_Name_1                          GDC.Case.id\n",
       "  Model_Data_Element_Name_2                    ICDC.case.case_id\n",
       "  Model_Data_Element_Name_3  IDC.DICOM.Patient Module.Patient ID\n",
       "  Model_Data_Element_Name_4                     PDC.Case.case_id"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models_stack.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get rid of \"model data element name\" column\n",
    "df_models_stack = df_models_stack.droplevel(-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDC.Case.id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ICDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDC.DICOM.Patient Module.Patient ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PDC.Case.case_id</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     0\n",
       "0                    CTDC.case.case_id\n",
       "0                          GDC.Case.id\n",
       "0                    ICDC.case.case_id\n",
       "0  IDC.DICOM.Patient Module.Patient ID\n",
       "0                     PDC.Case.case_id"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models_stack.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: index here aligns with index of DSS property"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2290"
      ]
     },
     "execution_count": 104,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_models_stack)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Clean stacked models df\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Drop any missing rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop any missing rows\n",
    "df_models_stack.dropna(subset=[0], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2290"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_models_stack)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "No missing values found here, but some incorrectly coded"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Replace non-breaking spaces and empty strings with null and drop nulls again"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_models_stack[0].replace(u\"\\xa0\", np.nan, inplace=True)\n",
    "df_models_stack[0].replace(u\"\", np.nan, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_models_stack.dropna(subset=[0], inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2257"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_models_stack)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "33 empty rows dropped this time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CTDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GDC.Case.id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ICDC.case.case_id</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDC.DICOM.Patient Module.Patient ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>PDC.Case.case_id</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     0\n",
       "0                    CTDC.case.case_id\n",
       "0                          GDC.Case.id\n",
       "0                    ICDC.case.case_id\n",
       "0  IDC.DICOM.Patient Module.Patient ID\n",
       "0                     PDC.Case.case_id"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models_stack.head()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Separate stacked models df into model, node, & prop handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_models_split = df_models_stack[0].str.split(\".\", expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    2257\n",
       "1    2257\n",
       "2    2227\n",
       "3      15\n",
       "dtype: int64"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models_split.count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Sex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>General Series Module</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>271</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Study Module</td>\n",
       "      <td>Patient Age</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>517</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Study Module</td>\n",
       "      <td>Patient's Size</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>518</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Study Module</td>\n",
       "      <td>Patient's Weight</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>519</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>General Series Module</td>\n",
       "      <td>StudyDate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>571</th>\n",
       "      <td>GDC</td>\n",
       "      <td>SlideImage</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668</th>\n",
       "      <td>GDC</td>\n",
       "      <td>ClinicalSupplement</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>736</th>\n",
       "      <td>GDC</td>\n",
       "      <td>ExperimentMetadata</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>824</th>\n",
       "      <td>GDC</td>\n",
       "      <td>PathologyReport</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>894</th>\n",
       "      <td>GDC</td>\n",
       "      <td>ProteinExpression</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>940</th>\n",
       "      <td>GDC</td>\n",
       "      <td>RunMetadata</td>\n",
       "      <td>ref:GDC</td>\n",
       "      <td>data_file_properties</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       0                   1                      2                       3\n",
       "0    IDC               DICOM         Patient Module              Patient ID\n",
       "2    IDC               DICOM         Patient Module  Patient's Ethnic Group\n",
       "10   IDC               DICOM         Patient Module  Patient's Ethnic Group\n",
       "29   IDC               DICOM         Patient Module           Patient's Sex\n",
       "48   IDC               DICOM  General Series Module        BodyPartExamined\n",
       "271  IDC               DICOM   Patient Study Module             Patient Age\n",
       "517  IDC               DICOM   Patient Study Module          Patient's Size\n",
       "518  IDC               DICOM   Patient Study Module        Patient's Weight\n",
       "519  IDC               DICOM  General Series Module               StudyDate\n",
       "571  GDC          SlideImage                ref:GDC    data_file_properties\n",
       "668  GDC  ClinicalSupplement                ref:GDC    data_file_properties\n",
       "736  GDC  ExperimentMetadata                ref:GDC    data_file_properties\n",
       "824  GDC     PathologyReport                ref:GDC    data_file_properties\n",
       "894  GDC   ProteinExpression                ref:GDC    data_file_properties\n",
       "940  GDC         RunMetadata                ref:GDC    data_file_properties"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_models rows with a third option (what should be node and what should be property?)\n",
    "df_models_other_1 = df_models_split.loc[df_models_split[3].notnull()]\n",
    "df_models_other_1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>50</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>case</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>diagnosis</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>53</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>sample</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>demographic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>program</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>study</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>file</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>principal_investigator</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>enrollment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>212</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>sample</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>260</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>adverse_event</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>261</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>follow_up</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>404</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>disease_extent</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>GDC desc</td>\n",
       "      <td>ENTITY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426</th>\n",
       "      <td>PDC desc</td>\n",
       "      <td>ENTITY</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>prior_therapy</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>498</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>study_arm</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>594</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>agent</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>598</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>agent_administration</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>648</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>assay</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>671</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>cohort</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>675</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>cycle</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>765</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>image</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>802</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>off_study</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>812</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>off_treatment</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>827</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>physical_exam</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>prior_surgery</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>999</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>study_site</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1011</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>visit</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1015</th>\n",
       "      <td>ICDC</td>\n",
       "      <td>vital_signs</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             0                       1     2     3\n",
       "50        ICDC                    case  None  None\n",
       "51        ICDC               diagnosis  None  None\n",
       "53        ICDC                  sample  None  None\n",
       "55        ICDC             demographic  None  None\n",
       "56        ICDC                 program  None  None\n",
       "57        ICDC                   study  None  None\n",
       "75        ICDC                    file  None  None\n",
       "79        ICDC  principal_investigator  None  None\n",
       "134       ICDC              enrollment  None  None\n",
       "212       ICDC                  sample  None  None\n",
       "260       ICDC           adverse_event  None  None\n",
       "261       ICDC               follow_up  None  None\n",
       "404       ICDC          disease_extent  None  None\n",
       "426   GDC desc                  ENTITY  None  None\n",
       "426   PDC desc                  ENTITY  None  None\n",
       "472       ICDC           prior_therapy  None  None\n",
       "498       ICDC               study_arm  None  None\n",
       "594       ICDC                   agent  None  None\n",
       "598       ICDC    agent_administration  None  None\n",
       "648       ICDC                   assay  None  None\n",
       "671       ICDC                  cohort  None  None\n",
       "675       ICDC                   cycle  None  None\n",
       "765       ICDC                   image  None  None\n",
       "802       ICDC               off_study  None  None\n",
       "812       ICDC           off_treatment  None  None\n",
       "827       ICDC           physical_exam  None  None\n",
       "842       ICDC           prior_surgery  None  None\n",
       "999       ICDC              study_site  None  None\n",
       "1011      ICDC                   visit  None  None\n",
       "1015      ICDC             vital_signs  None  None"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# df_models without a 2nd option\n",
    "df_models_other_2 = df_models_split.loc[df_models_split[1].notnull() & df_models_split[2].isnull()]\n",
    "df_models_other_2"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Identify duplicates in split models df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Specimen Profile</td>\n",
       "      <td>Type</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CTDC</td>\n",
       "      <td>case</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>IDC</td>\n",
       "      <td>DICOM</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PDC</td>\n",
       "      <td>Demographic</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CDA</td>\n",
       "      <td>Patient</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1425</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "      <td>Code</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1426</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "      <td>Body Site</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1430</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "      <td>Stage</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1502</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Tumor Size Profile</td>\n",
       "      <td>Component &gt; Tumor Longest Dimension</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1507</th>\n",
       "      <td>mCODE</td>\n",
       "      <td>Tumor Specimen Profile</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0                                 1  \\\n",
       "1     mCODE          Genomic Specimen Profile   \n",
       "2      CTDC                              case   \n",
       "2       IDC                             DICOM   \n",
       "2       PDC                       Demographic   \n",
       "2       CDA                           Patient   \n",
       "...     ...                               ...   \n",
       "1425  mCODE  Primary Cancer Condition Profile   \n",
       "1426  mCODE  Primary Cancer Condition Profile   \n",
       "1430  mCODE  Primary Cancer Condition Profile   \n",
       "1502  mCODE                Tumor Size Profile   \n",
       "1507  mCODE            Tumor Specimen Profile   \n",
       "\n",
       "                                        2                       3  \n",
       "1                                    Type                    None  \n",
       "2                               ethnicity                    None  \n",
       "2                          Patient Module  Patient's Ethnic Group  \n",
       "2                               ethnicity                    None  \n",
       "2                               ethnicity                    None  \n",
       "...                                   ...                     ...  \n",
       "1425                                 Code                    None  \n",
       "1426                            Body Site                    None  \n",
       "1430                                Stage                    None  \n",
       "1502  Component > Tumor Longest Dimension                    None  \n",
       "1507                           Identifier                    None  \n",
       "\n",
       "[106 rows x 4 columns]"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_models_split[df_models_split.duplicated(keep=False)]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Change model df handles to match mapping script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_models_split.columns = [\"ent_2_model\", \"ent_2_extra_handles\", \"ent_2_handle\", \"ent_2_handles_other\"]\n",
    "df_models_split = df_models_split[[\"ent_2_model\", \"ent_2_handle\", \"ent_2_handles_other\", \"ent_2_extra_handles\"]]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Combine DST and other model mappings into one df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mappings = df_dss.join(df_models_split)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC</td>\n",
       "      <td>gleason_grade_tertiary</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PDC</td>\n",
       "      <td>figo_staging_edition_year</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Exposure</td>\n",
       "      <td>PDC</td>\n",
       "      <td>exposure_type</td>\n",
       "      <td>None</td>\n",
       "      <td>Exposure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2266 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model    ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0            DSS      Identifier             Subject        CTDC   \n",
       "0            DSS      Identifier             Subject         GDC   \n",
       "0            DSS      Identifier             Subject        ICDC   \n",
       "0            DSS      Identifier             Subject         IDC   \n",
       "0            DSS      Identifier             Subject         PDC   \n",
       "...          ...             ...                 ...         ...   \n",
       "1525         DSS           Grade               Tumor         PDC   \n",
       "1526         DSS             NaN                 NaN         GDC   \n",
       "1526         DSS             NaN                 NaN         PDC   \n",
       "1527         DSS  Clinical Stage             Disease         PDC   \n",
       "1528         DSS            Type            Exposure         PDC   \n",
       "\n",
       "                   ent_2_handle ent_2_handles_other ent_2_extra_handles  \n",
       "0                       case_id                None                case  \n",
       "0                            id                None                Case  \n",
       "0                       case_id                None                case  \n",
       "0                Patient Module          Patient ID               DICOM  \n",
       "0                       case_id                None                Case  \n",
       "...                         ...                 ...                 ...  \n",
       "1525     gleason_grade_tertiary                None           Diagnosis  \n",
       "1526                tumor_depth                None           Diagnosis  \n",
       "1526                tumor_depth                None           Diagnosis  \n",
       "1527  figo_staging_edition_year                None           Diagnosis  \n",
       "1528              exposure_type                None            Exposure  \n",
       "\n",
       "[2266 rows x 7 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mappings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### identify duplicate rows in combined df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>PDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>Demographic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>961</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>of_case(case)</td>\n",
       "      <td>None</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>962</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>963</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1069</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1303</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>75 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model ent_1_handle ent_1_extra_handles ent_2_model    ent_2_handle  \\\n",
       "2            DSS   Ethnicity*             Subject        CTDC       ethnicity   \n",
       "2            DSS   Ethnicity*             Subject         IDC  Patient Module   \n",
       "2            DSS   Ethnicity*             Subject         PDC       ethnicity   \n",
       "2            DSS   Ethnicity*             Subject         CDA       ethnicity   \n",
       "10           DSS   Ethnicity*             Subject        CTDC       ethnicity   \n",
       "...          ...          ...                 ...         ...             ...   \n",
       "961          DSS          NaN                 NaN        ICDC   of_case(case)   \n",
       "962          DSS          NaN                 NaN         NaN             NaN   \n",
       "963          DSS          NaN                 NaN         NaN             NaN   \n",
       "1069         DSS          NaN                 NaN         NaN             NaN   \n",
       "1303         DSS          NaN                 NaN         NaN             NaN   \n",
       "\n",
       "         ent_2_handles_other ent_2_extra_handles  \n",
       "2                       None                case  \n",
       "2     Patient's Ethnic Group               DICOM  \n",
       "2                       None         Demographic  \n",
       "2                       None             Patient  \n",
       "10                      None                case  \n",
       "...                      ...                 ...  \n",
       "961                     None              sample  \n",
       "962                      NaN                 NaN  \n",
       "963                      NaN                 NaN  \n",
       "1069                     NaN                 NaN  \n",
       "1303                     NaN                 NaN  \n",
       "\n",
       "[75 rows x 7 columns]"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mappings[df_mappings.duplicated(keep=False)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2266"
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_mappings)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### drop duplicate rows in combined df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_mappings.drop_duplicates(keep=\"first\", inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2216"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_mappings)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "50 duplicate rows dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC</td>\n",
       "      <td>gleason_grade_tertiary</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PDC</td>\n",
       "      <td>figo_staging_edition_year</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Exposure</td>\n",
       "      <td>PDC</td>\n",
       "      <td>exposure_type</td>\n",
       "      <td>None</td>\n",
       "      <td>Exposure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2216 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model    ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0            DSS      Identifier             Subject        CTDC   \n",
       "0            DSS      Identifier             Subject         GDC   \n",
       "0            DSS      Identifier             Subject        ICDC   \n",
       "0            DSS      Identifier             Subject         IDC   \n",
       "0            DSS      Identifier             Subject         PDC   \n",
       "...          ...             ...                 ...         ...   \n",
       "1525         DSS           Grade               Tumor         PDC   \n",
       "1526         DSS             NaN                 NaN         GDC   \n",
       "1526         DSS             NaN                 NaN         PDC   \n",
       "1527         DSS  Clinical Stage             Disease         PDC   \n",
       "1528         DSS            Type            Exposure         PDC   \n",
       "\n",
       "                   ent_2_handle ent_2_handles_other ent_2_extra_handles  \n",
       "0                       case_id                None                case  \n",
       "0                            id                None                Case  \n",
       "0                       case_id                None                case  \n",
       "0                Patient Module          Patient ID               DICOM  \n",
       "0                       case_id                None                Case  \n",
       "...                         ...                 ...                 ...  \n",
       "1525     gleason_grade_tertiary                None           Diagnosis  \n",
       "1526                tumor_depth                None           Diagnosis  \n",
       "1526                tumor_depth                None           Diagnosis  \n",
       "1527  figo_staging_edition_year                None           Diagnosis  \n",
       "1528              exposure_type                None            Exposure  \n",
       "\n",
       "[2216 rows x 7 columns]"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_mappings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Separate rows with missing DSS property handle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings_other = df_mappings[df_mappings[\"ent_1_handle\"].isna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC</td>\n",
       "      <td>site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>IDC</td>\n",
       "      <td>tcia_tumorLocation</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDC</td>\n",
       "      <td>site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>CDS</td>\n",
       "      <td>site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>Additional Diagnosis Information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HTAN</td>\n",
       "      <td>site_of_resection_or_biopsy</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1520</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Location</td>\n",
       "      <td>None</td>\n",
       "      <td>Tumor Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1521</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Location Qualifier</td>\n",
       "      <td>None</td>\n",
       "      <td>Tumor Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1522</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Patient</td>\n",
       "      <td>None</td>\n",
       "      <td>Tumor Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1526</th>\n",
       "      <td>DSS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>PDC</td>\n",
       "      <td>tumor_depth</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1904 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "4            DSS          NaN                 NaN         GDC   \n",
       "4            DSS          NaN                 NaN         IDC   \n",
       "4            DSS          NaN                 NaN         PDC   \n",
       "4            DSS          NaN                 NaN         CDS   \n",
       "4            DSS          NaN                 NaN        HTAN   \n",
       "...          ...          ...                 ...         ...   \n",
       "1520         DSS          NaN                 NaN       mCODE   \n",
       "1521         DSS          NaN                 NaN       mCODE   \n",
       "1522         DSS          NaN                 NaN       mCODE   \n",
       "1526         DSS          NaN                 NaN         GDC   \n",
       "1526         DSS          NaN                 NaN         PDC   \n",
       "\n",
       "                     ent_2_handle ent_2_handles_other  \\\n",
       "4     site_of_resection_or_biopsy                None   \n",
       "4              tcia_tumorLocation                None   \n",
       "4     site_of_resection_or_biopsy                None   \n",
       "4     site_of_resection_or_biopsy                None   \n",
       "4     site_of_resection_or_biopsy                None   \n",
       "...                           ...                 ...   \n",
       "1520                     Location                None   \n",
       "1521           Location Qualifier                None   \n",
       "1522                      Patient                None   \n",
       "1526                  tumor_depth                None   \n",
       "1526                  tumor_depth                None   \n",
       "\n",
       "                   ent_2_extra_handles  \n",
       "4                            Diagnosis  \n",
       "4                                 TCIA  \n",
       "4                            Diagnosis  \n",
       "4     Additional Diagnosis Information  \n",
       "4                            Diagnosis  \n",
       "...                                ...  \n",
       "1520                     Tumor Profile  \n",
       "1521                     Tumor Profile  \n",
       "1522                     Tumor Profile  \n",
       "1526                         Diagnosis  \n",
       "1526                         Diagnosis  \n",
       "\n",
       "[1904 rows x 7 columns]"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings_other"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings = df_mappings[df_mappings[\"ent_1_handle\"].notna()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>None</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1523</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC</td>\n",
       "      <td>who_cns_grade</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1524</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC</td>\n",
       "      <td>who_nte_grade</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1525</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>PDC</td>\n",
       "      <td>gleason_grade_tertiary</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>PDC</td>\n",
       "      <td>figo_staging_edition_year</td>\n",
       "      <td>None</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1528</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Exposure</td>\n",
       "      <td>PDC</td>\n",
       "      <td>exposure_type</td>\n",
       "      <td>None</td>\n",
       "      <td>Exposure</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>312 rows × 7 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     ent_1_model    ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0            DSS      Identifier             Subject        CTDC   \n",
       "0            DSS      Identifier             Subject         GDC   \n",
       "0            DSS      Identifier             Subject        ICDC   \n",
       "0            DSS      Identifier             Subject         IDC   \n",
       "0            DSS      Identifier             Subject         PDC   \n",
       "...          ...             ...                 ...         ...   \n",
       "1523         DSS           Grade               Tumor         PDC   \n",
       "1524         DSS           Grade               Tumor         PDC   \n",
       "1525         DSS           Grade               Tumor         PDC   \n",
       "1527         DSS  Clinical Stage             Disease         PDC   \n",
       "1528         DSS            Type            Exposure         PDC   \n",
       "\n",
       "                   ent_2_handle ent_2_handles_other ent_2_extra_handles  \n",
       "0                       case_id                None                case  \n",
       "0                            id                None                Case  \n",
       "0                       case_id                None                case  \n",
       "0                Patient Module          Patient ID               DICOM  \n",
       "0                       case_id                None                Case  \n",
       "...                         ...                 ...                 ...  \n",
       "1523              who_cns_grade                None           Diagnosis  \n",
       "1524              who_nte_grade                None           Diagnosis  \n",
       "1525     gleason_grade_tertiary                None           Diagnosis  \n",
       "1527  figo_staging_edition_year                None           Diagnosis  \n",
       "1528              exposure_type                None            Exposure  \n",
       "\n",
       "[312 rows x 7 columns]"
      ]
     },
     "execution_count": 127,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Fix identified issues with mappings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Clean IDC node handles; matching to CDA definition, TBD exactly how to do in MDB as of Dec 2022"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>IDC</td>\n",
       "      <td>source_DOI</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Name</td>\n",
       "      <td>Program</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Program</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>tcia_species (not yet available in IDC product...</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>IDC</td>\n",
       "      <td>General Series Module</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>IDC</td>\n",
       "      <td>collection_id</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Modality</td>\n",
       "      <td>None</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0           DSS                Identifier             Subject         IDC   \n",
       "2           DSS                Ethnicity*             Subject         IDC   \n",
       "14          DSS                Identifier                File         IDC   \n",
       "15          DSS                      Name             Program         IDC   \n",
       "21          DSS                     Taxon             Subject         IDC   \n",
       "29          DSS  Gender (Gender Identity)             Subject         IDC   \n",
       "48          DSS             Anatomic Site            Specimen         IDC   \n",
       "259         DSS                Identifier             Project         IDC   \n",
       "277         DSS                  Modality             Imaging         IDC   \n",
       "\n",
       "                                          ent_2_handle  \\\n",
       "0                                       Patient Module   \n",
       "2                                       Patient Module   \n",
       "14                                          source_DOI   \n",
       "15                                             Program   \n",
       "21   tcia_species (not yet available in IDC product...   \n",
       "29                                      Patient Module   \n",
       "48                               General Series Module   \n",
       "259                                      collection_id   \n",
       "277                                           Modality   \n",
       "\n",
       "        ent_2_handles_other ent_2_extra_handles  \n",
       "0                Patient ID               DICOM  \n",
       "2    Patient's Ethnic Group               DICOM  \n",
       "14                     None                TCIA  \n",
       "15                     None                TCIA  \n",
       "21                     None                TCIA  \n",
       "29            Patient's Sex               DICOM  \n",
       "48         BodyPartExamined               DICOM  \n",
       "259                    None                TCIA  \n",
       "277                    None               DICOM  "
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# giving IDC nodes the \"file\" node for now to match existing (from CDA?)\n",
    "# could change to use tables from https://learn.canceridc.dev/data/organization-of-data/files-and-metadata as nodes?\n",
    "\n",
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"IDC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient Module</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>IDC</td>\n",
       "      <td>General Series Module</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0          DSS                Identifier             Subject         IDC   \n",
       "2          DSS                Ethnicity*             Subject         IDC   \n",
       "29         DSS  Gender (Gender Identity)             Subject         IDC   \n",
       "48         DSS             Anatomic Site            Specimen         IDC   \n",
       "\n",
       "             ent_2_handle     ent_2_handles_other ent_2_extra_handles  \n",
       "0          Patient Module              Patient ID               DICOM  \n",
       "2          Patient Module  Patient's Ethnic Group               DICOM  \n",
       "29         Patient Module           Patient's Sex               DICOM  \n",
       "48  General Series Module        BodyPartExamined               DICOM  "
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_handles_other\"].notnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "c:\\ProgramData\\Anaconda3\\envs\\bento_test_3_8\\lib\\site-packages\\pandas\\core\\indexing.py:1732: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._setitem_single_block(indexer, value, name)\n",
      "c:\\ProgramData\\Anaconda3\\envs\\bento_test_3_8\\lib\\site-packages\\pandas\\core\\indexing.py:723: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  iloc._setitem_with_indexer(indexer, value, self.name)\n"
     ]
    }
   ],
   "source": [
    "idc_other_handles_mask = dss_mappings[\"ent_2_handles_other\"].notnull()\n",
    "idc_other_handles = dss_mappings[dss_mappings[\"ent_2_handles_other\"].notnull()][\"ent_2_handles_other\"]\n",
    "dss_mappings.loc[idc_other_handles_mask, 'ent_2_handle'] = idc_other_handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_handles_other</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>IDC</td>\n",
       "      <td>source_DOI</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Name</td>\n",
       "      <td>Program</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Program</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>tcia_species (not yet available in IDC product...</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>IDC</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>IDC</td>\n",
       "      <td>collection_id</td>\n",
       "      <td>None</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Modality</td>\n",
       "      <td>None</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0           DSS                Identifier             Subject         IDC   \n",
       "2           DSS                Ethnicity*             Subject         IDC   \n",
       "14          DSS                Identifier                File         IDC   \n",
       "15          DSS                      Name             Program         IDC   \n",
       "21          DSS                     Taxon             Subject         IDC   \n",
       "29          DSS  Gender (Gender Identity)             Subject         IDC   \n",
       "48          DSS             Anatomic Site            Specimen         IDC   \n",
       "259         DSS                Identifier             Project         IDC   \n",
       "277         DSS                  Modality             Imaging         IDC   \n",
       "\n",
       "                                          ent_2_handle  \\\n",
       "0                                           Patient ID   \n",
       "2                               Patient's Ethnic Group   \n",
       "14                                          source_DOI   \n",
       "15                                             Program   \n",
       "21   tcia_species (not yet available in IDC product...   \n",
       "29                                       Patient's Sex   \n",
       "48                                    BodyPartExamined   \n",
       "259                                      collection_id   \n",
       "277                                           Modality   \n",
       "\n",
       "        ent_2_handles_other ent_2_extra_handles  \n",
       "0                Patient ID               DICOM  \n",
       "2    Patient's Ethnic Group               DICOM  \n",
       "14                     None                TCIA  \n",
       "15                     None                TCIA  \n",
       "21                     None                TCIA  \n",
       "29            Patient's Sex               DICOM  \n",
       "48         BodyPartExamined               DICOM  \n",
       "259                    None                TCIA  \n",
       "277                    None               DICOM  "
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"IDC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings = dss_mappings.drop(columns=\"ent_2_handles_other\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>IDC</td>\n",
       "      <td>source_DOI</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Name</td>\n",
       "      <td>Program</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Program</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>tcia_species (not yet available in IDC product...</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>IDC</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>IDC</td>\n",
       "      <td>collection_id</td>\n",
       "      <td>TCIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Modality</td>\n",
       "      <td>DICOM</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0           DSS                Identifier             Subject         IDC   \n",
       "2           DSS                Ethnicity*             Subject         IDC   \n",
       "14          DSS                Identifier                File         IDC   \n",
       "15          DSS                      Name             Program         IDC   \n",
       "21          DSS                     Taxon             Subject         IDC   \n",
       "29          DSS  Gender (Gender Identity)             Subject         IDC   \n",
       "48          DSS             Anatomic Site            Specimen         IDC   \n",
       "259         DSS                Identifier             Project         IDC   \n",
       "277         DSS                  Modality             Imaging         IDC   \n",
       "\n",
       "                                          ent_2_handle ent_2_extra_handles  \n",
       "0                                           Patient ID               DICOM  \n",
       "2                               Patient's Ethnic Group               DICOM  \n",
       "14                                          source_DOI                TCIA  \n",
       "15                                             Program                TCIA  \n",
       "21   tcia_species (not yet available in IDC product...                TCIA  \n",
       "29                                       Patient's Sex               DICOM  \n",
       "48                                    BodyPartExamined               DICOM  \n",
       "259                                      collection_id                TCIA  \n",
       "277                                           Modality               DICOM  "
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"IDC\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings.loc[dss_mappings[\"ent_2_model\"] == \"IDC\", [\"ent_2_extra_handles\"]] = \"file\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity*</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Ethnic Group</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>IDC</td>\n",
       "      <td>source_DOI</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Name</td>\n",
       "      <td>Program</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Program</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>tcia_species (not yet available in IDC product...</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient's Sex</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>IDC</td>\n",
       "      <td>BodyPartExamined</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>259</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>IDC</td>\n",
       "      <td>collection_id</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>277</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Modality</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0           DSS                Identifier             Subject         IDC   \n",
       "2           DSS                Ethnicity*             Subject         IDC   \n",
       "14          DSS                Identifier                File         IDC   \n",
       "15          DSS                      Name             Program         IDC   \n",
       "21          DSS                     Taxon             Subject         IDC   \n",
       "29          DSS  Gender (Gender Identity)             Subject         IDC   \n",
       "48          DSS             Anatomic Site            Specimen         IDC   \n",
       "259         DSS                Identifier             Project         IDC   \n",
       "277         DSS                  Modality             Imaging         IDC   \n",
       "\n",
       "                                          ent_2_handle ent_2_extra_handles  \n",
       "0                                           Patient ID                file  \n",
       "2                               Patient's Ethnic Group                file  \n",
       "14                                          source_DOI                file  \n",
       "15                                             Program                file  \n",
       "21   tcia_species (not yet available in IDC product...                file  \n",
       "29                                       Patient's Sex                file  \n",
       "48                                    BodyPartExamined                file  \n",
       "259                                      collection_id                file  \n",
       "277                                           Modality                file  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"IDC\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Clean up other notes, etc. added to handles in raw spreadsheet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nelso\\AppData\\Local\\Temp\\ipykernel_25752\\1100968097.py:7: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  dss_mappings[\"ent_1_handle\"] = dss_mappings[\"ent_1_handle\"].str.replace(ent_1_pattern, \"\")\n"
     ]
    }
   ],
   "source": [
    "# remove extraneous notes, etc. from ent 1 handles\n",
    "ent_1_pattern = \"|\".join([\n",
    "    re.escape(\"*\"),\n",
    "    re.escape(\" (same as name?)\"),\n",
    "    re.escape(\" (?)\")\n",
    "])\n",
    "dss_mappings[\"ent_1_handle\"] = dss_mappings[\"ent_1_handle\"].str.replace(ent_1_pattern, \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nelso\\AppData\\Local\\Temp\\ipykernel_25752\\4100785705.py:10: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  dss_mappings[\"ent_2_handle\"] = dss_mappings[\"ent_2_handle\"].str.replace(ent_2_pattern, \"\")\n"
     ]
    }
   ],
   "source": [
    "# remove extraneous notes, etc. from ent 2 handles\n",
    "ent_2_pattern = \"|\".join([\n",
    "    re.escape(\" (not yet available in IDC production release)\"),\n",
    "    re.escape(\" (Consider PhenotypicSex)\"),\n",
    "    re.escape(\"dct:\"),\n",
    "    re.escape(\"Component > \"),\n",
    "    re.escape(\"Collection > \"),\n",
    "    re.escape(\"'\")\n",
    "])\n",
    "dss_mappings[\"ent_2_handle\"] = dss_mappings[\"ent_2_handle\"].str.replace(ent_2_pattern, \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {},
   "outputs": [],
   "source": [
    "# reset index\n",
    "dss_mappings.reset_index(inplace=True, drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>Case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDS</td>\n",
       "      <td>subject_id</td>\n",
       "      <td>Participant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>biosample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Cancer Patient Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>GDC</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>Sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>physical_sample_type</td>\n",
       "      <td>sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>PDC</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>Sample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDS</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>Sample Information</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>source_material_type</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>HTAN</td>\n",
       "      <td>biospecimen_type</td>\n",
       "      <td>Molecular Test</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>biosample</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Type</td>\n",
       "      <td>Genomic Specimen Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>case</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>GDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Demographic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patients Ethnic Group</td>\n",
       "      <td>file</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ent_1_model ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0          DSS   Identifier             Subject        CTDC   \n",
       "1          DSS   Identifier             Subject         GDC   \n",
       "2          DSS   Identifier             Subject        ICDC   \n",
       "3          DSS   Identifier             Subject         IDC   \n",
       "4          DSS   Identifier             Subject         PDC   \n",
       "5          DSS   Identifier             Subject         CDS   \n",
       "6          DSS   Identifier             Subject         CDA   \n",
       "7          DSS   Identifier             Subject        C2M2   \n",
       "8          DSS   Identifier             Subject       mCODE   \n",
       "9          DSS         Type            Specimen         GDC   \n",
       "10         DSS         Type            Specimen        ICDC   \n",
       "11         DSS         Type            Specimen         PDC   \n",
       "12         DSS         Type            Specimen         CDS   \n",
       "13         DSS         Type            Specimen         CDA   \n",
       "14         DSS         Type            Specimen        HTAN   \n",
       "15         DSS         Type            Specimen        C2M2   \n",
       "16         DSS         Type            Specimen       mCODE   \n",
       "17         DSS    Ethnicity             Subject        CTDC   \n",
       "18         DSS    Ethnicity             Subject         GDC   \n",
       "19         DSS    Ethnicity             Subject         IDC   \n",
       "\n",
       "             ent_2_handle       ent_2_extra_handles  \n",
       "0                 case_id                      case  \n",
       "1                      id                      Case  \n",
       "2                 case_id                      case  \n",
       "3              Patient ID                      file  \n",
       "4                 case_id                      Case  \n",
       "5              subject_id               Participant  \n",
       "6              identifier                   Patient  \n",
       "7                local_id                 biosample  \n",
       "8              Identifier    Cancer Patient Profile  \n",
       "9             sample_type                    Sample  \n",
       "10  physical_sample_type                     sample  \n",
       "11            sample_type                    Sample  \n",
       "12            sample_type        Sample Information  \n",
       "13   source_material_type                  Specimen  \n",
       "14       biospecimen_type            Molecular Test  \n",
       "15             assay_type                 biosample  \n",
       "16                   Type  Genomic Specimen Profile  \n",
       "17              ethnicity                      case  \n",
       "18              ethnicity               Demographic  \n",
       "19  Patients Ethnic Group                      file  "
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings.head(20)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Clean up mCODE handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Cancer Patient Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Type</td>\n",
       "      <td>Genomic Specimen Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Stage</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>general morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Histology Morphology Behavior</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Tumor Specimen Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Gender</td>\n",
       "      <td>Cancer Patient Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Body Site</td>\n",
       "      <td>Genomic Specimen Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Site</td>\n",
       "      <td>Primary</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Body Site</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>DSS</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Cytogenetic Location</td>\n",
       "      <td>Genomic Variant Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Locus</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>DSS</td>\n",
       "      <td>symbol</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>DSS</td>\n",
       "      <td>specific morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>DSS</td>\n",
       "      <td>longest dimension</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Tumor Longest Dimension</td>\n",
       "      <td>Tumor Size Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Pathologic Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Code</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Component</td>\n",
       "      <td>Comorbidities Elixhauser Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>DSS</td>\n",
       "      <td>symbol</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Cytogenetic Nomenclature</td>\n",
       "      <td>Genomic Variant Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>DSS</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied Profile</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Tumor Specimen Profile</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "8           DSS                Identifier             Subject       mCODE   \n",
       "16          DSS                      Type            Specimen       mCODE   \n",
       "37          DSS                     Grade               Tumor       mCODE   \n",
       "48          DSS       general morphology               Tissue       mCODE   \n",
       "60          DSS                Identifier            Specimen       mCODE   \n",
       "83          DSS  Gender (Gender Identity)             Subject       mCODE   \n",
       "96          DSS            Clinical Stage             Disease       mCODE   \n",
       "109         DSS             Anatomic Site            Specimen       mCODE   \n",
       "128         DSS                      Site             Primary       mCODE   \n",
       "144         DSS                chromosome                Gene       mCODE   \n",
       "146         DSS                     Locus                Gene       mCODE   \n",
       "148         DSS                    symbol                Gene       mCODE   \n",
       "162         DSS      specific morphology               Tissue       mCODE   \n",
       "174         DSS         longest dimension            Specimen       mCODE   \n",
       "182         DSS          Pathologic Stage             Disease       mCODE   \n",
       "192         DSS                      Code             Disease       mCODE   \n",
       "269         DSS                    symbol                Gene       mCODE   \n",
       "270         DSS                chromosome                Gene       mCODE   \n",
       "285         DSS                Identifier            Specimen       mCODE   \n",
       "\n",
       "                      ent_2_handle               ent_2_extra_handles  \n",
       "8                       Identifier            Cancer Patient Profile  \n",
       "16                            Type          Genomic Specimen Profile  \n",
       "37                           Stage  Primary Cancer Condition Profile  \n",
       "48   Histology Morphology Behavior  Primary Cancer Condition Profile  \n",
       "60                      Identifier            Tumor Specimen Profile  \n",
       "83                          Gender            Cancer Patient Profile  \n",
       "96                            Code  Primary Cancer Condition Profile  \n",
       "109                      Body Site          Genomic Specimen Profile  \n",
       "128                      Body Site  Primary Cancer Condition Profile  \n",
       "144           Cytogenetic Location           Genomic Variant Profile  \n",
       "146  Genomic Reference Sequence Id    Genomic Region Studied Profile  \n",
       "148  Genomic Reference Sequence Id    Genomic Region Studied Profile  \n",
       "162                           Code  Primary Cancer Condition Profile  \n",
       "174        Tumor Longest Dimension                Tumor Size Profile  \n",
       "182                           Code  Primary Cancer Condition Profile  \n",
       "192                      Component  Comorbidities Elixhauser Profile  \n",
       "269       Cytogenetic Nomenclature           Genomic Variant Profile  \n",
       "270  Genomic Reference Sequence Id    Genomic Region Studied Profile  \n",
       "285                     Identifier            Tumor Specimen Profile  "
      ]
     },
     "execution_count": 141,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"mCODE\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\nelso\\AppData\\Local\\Temp\\ipykernel_25752\\2338541350.py:5: FutureWarning: The default value of regex will change from True to False in a future version.\n",
      "  dss_mappings[\"ent_2_extra_handles\"] = dss_mappings[\"ent_2_extra_handles\"].str.replace(ent_2_extra_pattern, \"\")\n"
     ]
    }
   ],
   "source": [
    "# drop \"Profile\" from ent 2 node handles\n",
    "ent_2_extra_pattern = \"|\".join([\n",
    "    re.escape(\" Profile\"),\n",
    "])\n",
    "dss_mappings[\"ent_2_extra_handles\"] = dss_mappings[\"ent_2_extra_handles\"].str.replace(ent_2_extra_pattern, \"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Cancer Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Type</td>\n",
       "      <td>Genomic Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Grade</td>\n",
       "      <td>Tumor</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Stage</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>DSS</td>\n",
       "      <td>general morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Histology Morphology Behavior</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Tumor Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (Gender Identity)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Gender</td>\n",
       "      <td>Cancer Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Body Site</td>\n",
       "      <td>Genomic Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>128</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Site</td>\n",
       "      <td>Primary</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Body Site</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>144</th>\n",
       "      <td>DSS</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Cytogenetic Location</td>\n",
       "      <td>Genomic Variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>146</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Locus</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>148</th>\n",
       "      <td>DSS</td>\n",
       "      <td>symbol</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>162</th>\n",
       "      <td>DSS</td>\n",
       "      <td>specific morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>174</th>\n",
       "      <td>DSS</td>\n",
       "      <td>longest dimension</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Tumor Longest Dimension</td>\n",
       "      <td>Tumor Size</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>182</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Pathologic Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Code</td>\n",
       "      <td>Primary Cancer Condition</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>192</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Code</td>\n",
       "      <td>Disease</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Component</td>\n",
       "      <td>Comorbidities Elixhauser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>269</th>\n",
       "      <td>DSS</td>\n",
       "      <td>symbol</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Cytogenetic Nomenclature</td>\n",
       "      <td>Genomic Variant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>DSS</td>\n",
       "      <td>chromosome</td>\n",
       "      <td>Gene</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Genomic Reference Sequence Id</td>\n",
       "      <td>Genomic Region Studied</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Tumor Specimen</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model              ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "8           DSS                Identifier             Subject       mCODE   \n",
       "16          DSS                      Type            Specimen       mCODE   \n",
       "37          DSS                     Grade               Tumor       mCODE   \n",
       "48          DSS       general morphology               Tissue       mCODE   \n",
       "60          DSS                Identifier            Specimen       mCODE   \n",
       "83          DSS  Gender (Gender Identity)             Subject       mCODE   \n",
       "96          DSS            Clinical Stage             Disease       mCODE   \n",
       "109         DSS             Anatomic Site            Specimen       mCODE   \n",
       "128         DSS                      Site             Primary       mCODE   \n",
       "144         DSS                chromosome                Gene       mCODE   \n",
       "146         DSS                     Locus                Gene       mCODE   \n",
       "148         DSS                    symbol                Gene       mCODE   \n",
       "162         DSS      specific morphology               Tissue       mCODE   \n",
       "174         DSS         longest dimension            Specimen       mCODE   \n",
       "182         DSS          Pathologic Stage             Disease       mCODE   \n",
       "192         DSS                      Code             Disease       mCODE   \n",
       "269         DSS                    symbol                Gene       mCODE   \n",
       "270         DSS                chromosome                Gene       mCODE   \n",
       "285         DSS                Identifier            Specimen       mCODE   \n",
       "\n",
       "                      ent_2_handle       ent_2_extra_handles  \n",
       "8                       Identifier            Cancer Patient  \n",
       "16                            Type          Genomic Specimen  \n",
       "37                           Stage  Primary Cancer Condition  \n",
       "48   Histology Morphology Behavior  Primary Cancer Condition  \n",
       "60                      Identifier            Tumor Specimen  \n",
       "83                          Gender            Cancer Patient  \n",
       "96                            Code  Primary Cancer Condition  \n",
       "109                      Body Site          Genomic Specimen  \n",
       "128                      Body Site  Primary Cancer Condition  \n",
       "144           Cytogenetic Location           Genomic Variant  \n",
       "146  Genomic Reference Sequence Id    Genomic Region Studied  \n",
       "148  Genomic Reference Sequence Id    Genomic Region Studied  \n",
       "162                           Code  Primary Cancer Condition  \n",
       "174        Tumor Longest Dimension                Tumor Size  \n",
       "182                           Code  Primary Cancer Condition  \n",
       "192                      Component  Comorbidities Elixhauser  \n",
       "269       Cytogenetic Nomenclature           Genomic Variant  \n",
       "270  Genomic Reference Sequence Id    Genomic Region Studied  \n",
       "285                     Identifier            Tumor Specimen  "
      ]
     },
     "execution_count": 143,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"mCODE\"]"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Clean up CDA handles (to match those actually provided by CDA)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>source_material_type</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Race</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>race</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Age at Diagnosis</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>age_at_diagnosis</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>DSS</td>\n",
       "      <td>general morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>CDA</td>\n",
       "      <td>general_tissue_morphology</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Format</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>file_format</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>taxon</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>data_type</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>anatomical_site</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (PhenotypicSex)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>sex</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Sex (Genotypic Sex)\\n</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasGenotypicSex</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Investigator Name</td>\n",
       "      <td>Principal</td>\n",
       "      <td>CDA</td>\n",
       "      <td>isPrincipalInvestigator</td>\n",
       "      <td>Dataset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>DSS</td>\n",
       "      <td>specific morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>CDA</td>\n",
       "      <td>specific_tissue_morphology</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Project</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Pathologic Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasDiseaseStaging</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasDiseaseStaging</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Code</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasDisease?     [OntologyReference]</td>\n",
       "      <td>Patient</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasImagingFileType</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model            ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "6           DSS              Identifier             Subject         CDA   \n",
       "13          DSS                    Type            Specimen         CDA   \n",
       "22          DSS               Ethnicity             Subject         CDA   \n",
       "29          DSS                    Race             Subject         CDA   \n",
       "41          DSS        Age at Diagnosis             Subject         CDA   \n",
       "46          DSS     general morphology               Tissue         CDA   \n",
       "51          DSS                  Format                File         CDA   \n",
       "59          DSS              Identifier            Specimen         CDA   \n",
       "65          DSS              Identifier                File         CDA   \n",
       "75          DSS                   Taxon             Subject         CDA   \n",
       "87          DSS                    Type                File         CDA   \n",
       "107         DSS           Anatomic Site            Specimen         CDA   \n",
       "114         DSS  Gender (PhenotypicSex)             Subject         CDA   \n",
       "122         DSS   Sex (Genotypic Sex)\\n             Subject         CDA   \n",
       "133         DSS       Investigator Name           Principal         CDA   \n",
       "161         DSS    specific morphology               Tissue         CDA   \n",
       "168         DSS              Identifier             Project         CDA   \n",
       "180         DSS        Pathologic Stage             Disease         CDA   \n",
       "185         DSS          Clinical Stage             Disease         CDA   \n",
       "189         DSS                    Code             Disease         CDA   \n",
       "195         DSS                Modality             Imaging         CDA   \n",
       "\n",
       "                            ent_2_handle ent_2_extra_handles  \n",
       "6                             identifier             Patient  \n",
       "13                  source_material_type            Specimen  \n",
       "22                             ethnicity             Patient  \n",
       "29                                  race             Patient  \n",
       "41                      age_at_diagnosis           Diagnosis  \n",
       "46             general_tissue_morphology            Specimen  \n",
       "51                           file_format                File  \n",
       "59                            identifier            Specimen  \n",
       "65                            identifier                File  \n",
       "75                                 taxon             Patient  \n",
       "87                             data_type                File  \n",
       "107                      anatomical_site            Specimen  \n",
       "114                                  sex             Patient  \n",
       "122                      hasGenotypicSex             Patient  \n",
       "133              isPrincipalInvestigator             Dataset  \n",
       "161           specific_tissue_morphology            Specimen  \n",
       "168                           identifier             Project  \n",
       "180                    hasDiseaseStaging           Diagnosis  \n",
       "185                    hasDiseaseStaging           Diagnosis  \n",
       "189  hasDisease?     [OntologyReference]             Patient  \n",
       "195                   hasImagingFileType                File  "
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"CDA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDA \"Patient\" node to \"Subject\"\n",
    "cda_other_handles_mask_1 = (dss_mappings[\"ent_2_model\"] == \"CDA\") & (dss_mappings[\"ent_2_extra_handles\"] == \"Patient\")\n",
    "dss_mappings.loc[cda_other_handles_mask_1, \"ent_2_extra_handles\"] = \"Subject\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDA \"taxon\" property to \"species\"\n",
    "cda_other_handles_mask_2 = (dss_mappings[\"ent_2_model\"] == \"CDA\") & (dss_mappings[\"ent_2_handle\"] == \"taxon\")\n",
    "dss_mappings.loc[cda_other_handles_mask_2, \"ent_2_handle\"] = \"species\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDA \"hasDiseaseStaging\" property to \"stage\"\n",
    "cda_other_handles_mask_3 = (dss_mappings[\"ent_2_model\"] == \"CDA\") & (dss_mappings[\"ent_2_handle\"] == \"hasDiseaseStaging\")\n",
    "dss_mappings.loc[cda_other_handles_mask_3, \"ent_2_handle\"] = \"stage\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# CDA \"identifier\" property of \"Project\" node to \"member_of_research_project\" of \"ResearchSubjeect\"\n",
    "cda_other_handles_mask_4 = (dss_mappings[\"ent_2_model\"] == \"CDA\") & (dss_mappings[\"ent_2_extra_handles\"] == \"Project\") & (dss_mappings[\"ent_2_handle\"] == \"identifier\")\n",
    "dss_mappings.loc[cda_other_handles_mask_4, \"ent_2_handle\"] = \"member_of_research_project\"\n",
    "dss_mappings.loc[cda_other_handles_mask_4, \"ent_2_extra_handles\"] = \"ResearchSubject\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>source_material_type</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Race</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>race</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Age at Diagnosis</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>age_at_diagnosis</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>46</th>\n",
       "      <td>DSS</td>\n",
       "      <td>general morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>CDA</td>\n",
       "      <td>general_tissue_morphology</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Format</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>file_format</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>species</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>data_type</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>anatomical_site</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (PhenotypicSex)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>sex</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>122</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Sex (Genotypic Sex)\\n</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasGenotypicSex</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>133</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Investigator Name</td>\n",
       "      <td>Principal</td>\n",
       "      <td>CDA</td>\n",
       "      <td>isPrincipalInvestigator</td>\n",
       "      <td>Dataset</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>161</th>\n",
       "      <td>DSS</td>\n",
       "      <td>specific morphology</td>\n",
       "      <td>Tissue</td>\n",
       "      <td>CDA</td>\n",
       "      <td>specific_tissue_morphology</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>CDA</td>\n",
       "      <td>member_of_research_project</td>\n",
       "      <td>ResearchSubject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Pathologic Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>stage</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>stage</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>189</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Code</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasDisease?     [OntologyReference]</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasImagingFileType</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model            ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "6           DSS              Identifier             Subject         CDA   \n",
       "13          DSS                    Type            Specimen         CDA   \n",
       "22          DSS               Ethnicity             Subject         CDA   \n",
       "29          DSS                    Race             Subject         CDA   \n",
       "41          DSS        Age at Diagnosis             Subject         CDA   \n",
       "46          DSS     general morphology               Tissue         CDA   \n",
       "51          DSS                  Format                File         CDA   \n",
       "59          DSS              Identifier            Specimen         CDA   \n",
       "65          DSS              Identifier                File         CDA   \n",
       "75          DSS                   Taxon             Subject         CDA   \n",
       "87          DSS                    Type                File         CDA   \n",
       "107         DSS           Anatomic Site            Specimen         CDA   \n",
       "114         DSS  Gender (PhenotypicSex)             Subject         CDA   \n",
       "122         DSS   Sex (Genotypic Sex)\\n             Subject         CDA   \n",
       "133         DSS       Investigator Name           Principal         CDA   \n",
       "161         DSS    specific morphology               Tissue         CDA   \n",
       "168         DSS              Identifier             Project         CDA   \n",
       "180         DSS        Pathologic Stage             Disease         CDA   \n",
       "185         DSS          Clinical Stage             Disease         CDA   \n",
       "189         DSS                    Code             Disease         CDA   \n",
       "195         DSS                Modality             Imaging         CDA   \n",
       "\n",
       "                            ent_2_handle ent_2_extra_handles  \n",
       "6                             identifier             Subject  \n",
       "13                  source_material_type            Specimen  \n",
       "22                             ethnicity             Subject  \n",
       "29                                  race             Subject  \n",
       "41                      age_at_diagnosis           Diagnosis  \n",
       "46             general_tissue_morphology            Specimen  \n",
       "51                           file_format                File  \n",
       "59                            identifier            Specimen  \n",
       "65                            identifier                File  \n",
       "75                               species             Subject  \n",
       "87                             data_type                File  \n",
       "107                      anatomical_site            Specimen  \n",
       "114                                  sex             Subject  \n",
       "122                      hasGenotypicSex             Subject  \n",
       "133              isPrincipalInvestigator             Dataset  \n",
       "161           specific_tissue_morphology            Specimen  \n",
       "168           member_of_research_project     ResearchSubject  \n",
       "180                                stage           Diagnosis  \n",
       "185                                stage           Diagnosis  \n",
       "189  hasDisease?     [OntologyReference]             Subject  \n",
       "195                   hasImagingFileType                File  "
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"CDA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21, 6)"
      ]
     },
     "execution_count": 150,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"CDA\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop rows not found in official CDA mappings\n",
    "cda_drop_handles =  [\n",
    "    \"hasGenotypicSex\",\n",
    "    \"hasDisease?     [OntologyReference]\", \n",
    "    \"isPrincipalInvestigator\",\n",
    "    \"general_tissue_morphology\", \n",
    "    \"specific_tissue_morphology\"\n",
    "]\n",
    "\n",
    "for handle in cda_drop_handles:\n",
    "    cda_drop_index = dss_mappings.loc[(dss_mappings[\"ent_2_model\"] == \"CDA\") & (dss_mappings[\"ent_2_handle\"] == handle)].index\n",
    "    dss_mappings.drop(labels=cda_drop_index, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>source_material_type</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Race</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>race</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Age at Diagnosis</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>age_at_diagnosis</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Format</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>file_format</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Taxon</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>species</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>File</td>\n",
       "      <td>CDA</td>\n",
       "      <td>data_type</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Anatomic Site</td>\n",
       "      <td>Specimen</td>\n",
       "      <td>CDA</td>\n",
       "      <td>anatomical_site</td>\n",
       "      <td>Specimen</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>114</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Gender (PhenotypicSex)</td>\n",
       "      <td>Subject</td>\n",
       "      <td>CDA</td>\n",
       "      <td>sex</td>\n",
       "      <td>Subject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>Project</td>\n",
       "      <td>CDA</td>\n",
       "      <td>member_of_research_project</td>\n",
       "      <td>ResearchSubject</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>180</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Pathologic Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>stage</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>185</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Clinical Stage</td>\n",
       "      <td>Disease</td>\n",
       "      <td>CDA</td>\n",
       "      <td>stage</td>\n",
       "      <td>Diagnosis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>195</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Modality</td>\n",
       "      <td>Imaging</td>\n",
       "      <td>CDA</td>\n",
       "      <td>hasImagingFileType</td>\n",
       "      <td>File</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    ent_1_model            ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "6           DSS              Identifier             Subject         CDA   \n",
       "13          DSS                    Type            Specimen         CDA   \n",
       "22          DSS               Ethnicity             Subject         CDA   \n",
       "29          DSS                    Race             Subject         CDA   \n",
       "41          DSS        Age at Diagnosis             Subject         CDA   \n",
       "51          DSS                  Format                File         CDA   \n",
       "59          DSS              Identifier            Specimen         CDA   \n",
       "65          DSS              Identifier                File         CDA   \n",
       "75          DSS                   Taxon             Subject         CDA   \n",
       "87          DSS                    Type                File         CDA   \n",
       "107         DSS           Anatomic Site            Specimen         CDA   \n",
       "114         DSS  Gender (PhenotypicSex)             Subject         CDA   \n",
       "168         DSS              Identifier             Project         CDA   \n",
       "180         DSS        Pathologic Stage             Disease         CDA   \n",
       "185         DSS          Clinical Stage             Disease         CDA   \n",
       "195         DSS                Modality             Imaging         CDA   \n",
       "\n",
       "                   ent_2_handle ent_2_extra_handles  \n",
       "6                    identifier             Subject  \n",
       "13         source_material_type            Specimen  \n",
       "22                    ethnicity             Subject  \n",
       "29                         race             Subject  \n",
       "41             age_at_diagnosis           Diagnosis  \n",
       "51                  file_format                File  \n",
       "59                   identifier            Specimen  \n",
       "65                   identifier                File  \n",
       "75                      species             Subject  \n",
       "87                    data_type                File  \n",
       "107             anatomical_site            Specimen  \n",
       "114                         sex             Subject  \n",
       "168  member_of_research_project     ResearchSubject  \n",
       "180                       stage           Diagnosis  \n",
       "185                       stage           Diagnosis  \n",
       "195          hasImagingFileType                File  "
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"CDA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(16, 6)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings[dss_mappings[\"ent_2_model\"] == \"CDA\"].shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5 rows dropped b/c not in official CDA mappings"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Fix GDC, PDC, ICDC node handles to match those already in MDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CTDC', 'GDC', 'ICDC', 'IDC', 'PDC', 'CDS', 'CDA', 'C2M2', 'mCODE',\n",
       "       'HTAN', nan], dtype=object)"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings.ent_2_model.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert GDC node handles to lowercase to match format in MDB\n",
    "gdc_node_handles = dss_mappings[dss_mappings[\"ent_2_model\"] == \"GDC\"][\"ent_2_extra_handles\"]\n",
    "dss_mappings.loc[dss_mappings[\"ent_2_model\"] == \"GDC\", \"ent_2_extra_handles\"] = gdc_node_handles.str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert ICDC node handles to lowercase to match format in MDB\n",
    "icdc_node_handles = dss_mappings[dss_mappings[\"ent_2_model\"] == \"ICDC\"][\"ent_2_extra_handles\"]\n",
    "dss_mappings.loc[dss_mappings[\"ent_2_model\"] == \"ICDC\", \"ent_2_extra_handles\"] = icdc_node_handles.str.lower()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert PDC node handles to lowercase to match format in MDB\n",
    "pdc_node_handles = dss_mappings[dss_mappings[\"ent_2_model\"] == \"PDC\"][\"ent_2_extra_handles\"]\n",
    "dss_mappings.loc[dss_mappings[\"ent_2_model\"] == \"PDC\", \"ent_2_extra_handles\"] = pdc_node_handles.str.lower()"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Drop any rows with nulls or duplicates in cleaned df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(307, 6)"
      ]
     },
     "execution_count": 158,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Drop rows with a null"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(305, 6)"
      ]
     },
     "execution_count": 159,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop any rows with remaining NA values\n",
    "null_index = dss_mappings[dss_mappings.isnull().any(axis=1)].index\n",
    "dss_mappings.drop(labels=null_index, inplace=True)\n",
    "\n",
    "dss_mappings.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2 rows with null values dropped"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Drop duplicate rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 160,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(303, 6)"
      ]
     },
     "execution_count": 160,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# drop any remaining duplicates\n",
    "dss_mappings.drop_duplicates(keep=\"first\", inplace=True)\n",
    "dss_mappings.shape"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2 duplicate rows dropped"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Convert node handles to list for mapping script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 161,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings[\"ent_1_extra_handles\"] = [[l] for l in dss_mappings[\"ent_1_extra_handles\"]] # listify extra handles\n",
    "dss_mappings[\"ent_2_extra_handles\"] = [[l] for l in dss_mappings[\"ent_2_extra_handles\"]] # listify extra handles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ent_1_model</th>\n",
       "      <th>ent_1_handle</th>\n",
       "      <th>ent_1_extra_handles</th>\n",
       "      <th>ent_2_model</th>\n",
       "      <th>ent_2_handle</th>\n",
       "      <th>ent_2_extra_handles</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>[case]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>GDC</td>\n",
       "      <td>id</td>\n",
       "      <td>[case]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>[case]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patient ID</td>\n",
       "      <td>[file]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>PDC</td>\n",
       "      <td>case_id</td>\n",
       "      <td>[case]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>subject_id</td>\n",
       "      <td>[Participant]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>CDA</td>\n",
       "      <td>identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>local_id</td>\n",
       "      <td>[biosample]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Identifier</td>\n",
       "      <td>[Cancer Patient]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>GDC</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>[sample]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>ICDC</td>\n",
       "      <td>physical_sample_type</td>\n",
       "      <td>[sample]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>PDC</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>[sample]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>CDS</td>\n",
       "      <td>sample_type</td>\n",
       "      <td>[Sample Information]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>CDA</td>\n",
       "      <td>source_material_type</td>\n",
       "      <td>[Specimen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>HTAN</td>\n",
       "      <td>biospecimen_type</td>\n",
       "      <td>[Molecular Test]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>C2M2</td>\n",
       "      <td>assay_type</td>\n",
       "      <td>[biosample]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Specimen]</td>\n",
       "      <td>mCODE</td>\n",
       "      <td>Type</td>\n",
       "      <td>[Genomic Specimen]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>CTDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>[case]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>GDC</td>\n",
       "      <td>ethnicity</td>\n",
       "      <td>[demographic]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>DSS</td>\n",
       "      <td>Ethnicity</td>\n",
       "      <td>[Subject]</td>\n",
       "      <td>IDC</td>\n",
       "      <td>Patients Ethnic Group</td>\n",
       "      <td>[file]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ent_1_model ent_1_handle ent_1_extra_handles ent_2_model  \\\n",
       "0          DSS   Identifier           [Subject]        CTDC   \n",
       "1          DSS   Identifier           [Subject]         GDC   \n",
       "2          DSS   Identifier           [Subject]        ICDC   \n",
       "3          DSS   Identifier           [Subject]         IDC   \n",
       "4          DSS   Identifier           [Subject]         PDC   \n",
       "5          DSS   Identifier           [Subject]         CDS   \n",
       "6          DSS   Identifier           [Subject]         CDA   \n",
       "7          DSS   Identifier           [Subject]        C2M2   \n",
       "8          DSS   Identifier           [Subject]       mCODE   \n",
       "9          DSS         Type          [Specimen]         GDC   \n",
       "10         DSS         Type          [Specimen]        ICDC   \n",
       "11         DSS         Type          [Specimen]         PDC   \n",
       "12         DSS         Type          [Specimen]         CDS   \n",
       "13         DSS         Type          [Specimen]         CDA   \n",
       "14         DSS         Type          [Specimen]        HTAN   \n",
       "15         DSS         Type          [Specimen]        C2M2   \n",
       "16         DSS         Type          [Specimen]       mCODE   \n",
       "17         DSS    Ethnicity           [Subject]        CTDC   \n",
       "18         DSS    Ethnicity           [Subject]         GDC   \n",
       "19         DSS    Ethnicity           [Subject]         IDC   \n",
       "\n",
       "             ent_2_handle   ent_2_extra_handles  \n",
       "0                 case_id                [case]  \n",
       "1                      id                [case]  \n",
       "2                 case_id                [case]  \n",
       "3              Patient ID                [file]  \n",
       "4                 case_id                [case]  \n",
       "5              subject_id         [Participant]  \n",
       "6              identifier             [Subject]  \n",
       "7                local_id           [biosample]  \n",
       "8              Identifier      [Cancer Patient]  \n",
       "9             sample_type              [sample]  \n",
       "10  physical_sample_type               [sample]  \n",
       "11            sample_type              [sample]  \n",
       "12            sample_type  [Sample Information]  \n",
       "13   source_material_type            [Specimen]  \n",
       "14       biospecimen_type      [Molecular Test]  \n",
       "15             assay_type           [biosample]  \n",
       "16                   Type    [Genomic Specimen]  \n",
       "17              ethnicity                [case]  \n",
       "18              ethnicity         [demographic]  \n",
       "19  Patients Ethnic Group                [file]  "
      ]
     },
     "execution_count": 162,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dss_mappings.head(20)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output cleaned DSS mappings (& others that couldn't be mapped properly)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 178,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_path = Path(\"../data/output/\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 179,
   "metadata": {},
   "outputs": [],
   "source": [
    "today_date = date.today().strftime('%Y%m%d')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [],
   "source": [
    "clean_output_path = Path(output_path, f\"{today_date}_{input_filename}_clean.csv\")\n",
    "other_output_path = Path(output_path, f\"{today_date}_{input_filename}_other.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "metadata": {},
   "outputs": [],
   "source": [
    "dss_mappings.to_csv(clean_output_path, index=False)\n",
    "dss_mappings_other.to_csv(other_output_path, index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.8.13 ('bento_test_3_8')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.13"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "ce2a05d834a6e1455d54a8cfaa6704fb454d813a6ed075fd3c8a17f8880dd21f"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
